arduino/resources/index.go

```diff
@@ -0,0 +1,115 @@
+// To purchase a commercial license, send an email to license@arduino.cc.
+
+package resources
+
+import (
+	"net/url"
+	"path"
+	"strings"
+
+	"github.com/arduino/arduino-cli/arduino"
+	"github.com/arduino/arduino-cli/arduino/security"
+	"github.com/arduino/go-paths-helper"
+	"go.bug.st/downloader/v2"
+)
+
+// IndexResource is a reference to a package_index.json
+type IndexResource struct {
+	URL          *url.URL
+	SignatureURL *url.URL
+}
+
+// Download will download the index and possibly check the signature using the Arduino's public key.
+// If the file is in .gz format it will be unpacked first.
+func (res *IndexResource) Download(destDir *paths.Path, downloadCB DownloadProgressCB) error {
+	// Create destination directory
+	if err := destDir.MkdirAll(); err != nil {
+		return &arduino.PermissionDeniedError{Message: tr("Can't create data directory %s", destDir), Cause: err}
+	}
+
+	// Create a temp dir to stage all downloads
+	tmp, err := paths.MkTempDir("", "library_index_download")
+	if err != nil {
+		return &arduino.TempDirCreationFailedError{Cause: err}
+	}
+	defer tmp.RemoveAll()
+
+	// Download index file
+	indexFileName := path.Base(res.URL.Path) // == pakcage_index.json[.gz]
+	tmpIndexPath := tmp.Join(indexFileName)
+	if err := DownloadFile(tmpIndexPath, res.URL.String(), tr("Downloading index: %s", indexFileName), downloadCB, nil, downloader.NoResume); err != nil {
+		return &arduino.FailedDownloadError{Message: tr("Error downloading index '%s'", res.URL), Cause: err}
+	}
+
+	// Expand the index if it is compressed
+	if strings.HasSuffix(indexFileName, ".gz") {
+		indexFileName = strings.TrimSuffix(indexFileName, ".gz") // == pakcage_index.json
+		tmpUnzippedIndexPath := tmp.Join(indexFileName)
+		if err := paths.GUnzip(tmpIndexPath, tmpUnzippedIndexPath); err != nil {
+			return &arduino.PermissionDeniedError{Message: tr("Error extracting %s", indexFileName), Cause: err}
+		}
+		tmpIndexPath = tmpUnzippedIndexPath
+	}
+
+	// Check the signature if needed
+	var signaturePath, tmpSignaturePath *paths.Path
+	if res.SignatureURL != nil {
+		// Compose signature URL
+		signatureFileName := path.Base(res.SignatureURL.Path)
+
+		// Download signature
+		signaturePath = destDir.Join(signatureFileName)
+		tmpSignaturePath = tmp.Join(signatureFileName)
+		if err := DownloadFile(tmpSignaturePath, res.SignatureURL.String(), tr("Downloading index signature: %s", signatureFileName), downloadCB, nil, downloader.NoResume); err != nil {
+			return &arduino.FailedDownloadError{Message: tr("Error downloading index signature '%s'", res.SignatureURL), Cause: err}
+		}
+
+		// Check signature...
+		if valid, _, err := security.VerifyArduinoDetachedSignature(tmpIndexPath, tmpSignaturePath); err != nil {
+			return &arduino.PermissionDeniedError{Message: tr("Error verifying signature"), Cause: err}
+		} else if !valid {
+			return &arduino.SignatureVerificationFailedError{File: res.URL.String()}
+		}
+	}
+
+	// TODO: Implement a ResourceValidator
```

Comment:
What about this TODO? Will be done in another PR?

Reply:
Yes, previously we tried to parse the package_index.json with the packagemanager, but now we cannot do that anymore because we use the same subroutine to download also library_index (and potentially any other file since this function is generic enough).